framework/src/main/java/org/checkerframework/framework/util/QualifierKindHierarchy.java

```diff
@@ -0,0 +1,874 @@
+                }
+                if (qualifierKind.isPoly) {
+                    bot.superTypes.add(qualifierKind);
+                }
+            }
+        }
+    }
+
+    /**
+     * Returns the set of all qualifiers that are a supertype of {@code qualifierKind}.
+     *
+     * @param qualifierKind the qualifier whose super types should be returned
+     * @param directSuperMap directSuperMap
+     * @return the set of all qualifiers that are a supertype of {@code qualifierKind}
+     * @throws UserError if there is a cycle in the hierarchy
+     */
+    private Set<QualifierKind> findAllTheSupers(
+            QualifierKind qualifierKind, Map<QualifierKind, Set<QualifierKind>> directSuperMap) {
+        Queue<QualifierKind> queue = new ArrayDeque<>(directSuperMap.get(qualifierKind));
+        Set<QualifierKind> allSupers = new TreeSet<>(directSuperMap.get(qualifierKind));
+        while (!queue.isEmpty()) {
+            QualifierKind superQual = queue.remove();
+            if (superQual == qualifierKind) {
+                throw new UserError("Cycle in hierarchy: %s", qualifierKind);
+            }
+            queue.addAll(directSuperMap.get(superQual));
+            allSupers.addAll(directSuperMap.get(superQual));
+        }
+        return allSupers;
+    }
+
+    /**
+     * Creates a mapping from {@link QualifierKindPair} to the least upper bound of both
+     * QualifierKinds.
+     *
+     * @return a mapping from {@link QualifierKindPair} to their lub.
+     */
+    protected Map<QualifierKindPair, QualifierKind> createLubsMap() {
+        Map<QualifierKindPair, QualifierKind> lubs = new TreeMap<>();
+        for (QualifierKind qual1 : nameToQualifierKind.values()) {
+            for (QualifierKind qual2 : nameToQualifierKind.values()) {
+                if (qual1.top != qual2.top) {
+                    continue;
+                }
+                QualifierKind lub = findLub(qual1, qual2);
+                QualifierKindPair pair = new QualifierKindPair(qual1, qual2);
+                QualifierKind otherLub = lubs.get(pair);
+                if (otherLub != null) {
+                    if (otherLub != lub) {
+                        throw new BugInCF(
+                                "Multiple lubs for qualifiers %s and %s. Found lubs %s and %s",
+                                qual1, qual2, lub, otherLub);
+                    }
+                } else {
+                    lubs.put(pair, lub);
+                }
+            }
+        }
+        return lubs;
+    }
+
+    /**
+     * Returns the least upper bound of {@code qual1} and {@code qual2}.
+     *
+     * @param qual1 a qualifier kind
+     * @param qual2 a qualifier kind
+     * @return the least upper bound of {@code qual1} and {@code qual2}
+     */
+    private QualifierKind findLub(QualifierKind qual1, QualifierKind qual2) {
+        if (qual1 == qual2) {
+            return qual1;
+        } else if (qual1.isSubtype(qual2)) {
```

Comment:
This logic is a bit inefficient.  I think that's OK because this version is only ever called once per pair, at the beginning of the Checker Framework run, and the result is cached.  Explaining when various methods will be called would be useful to help readers understand the design.

Reply:
>Explaining when various methods will be called would be useful to help readers understand the design.

Any method in this class that does anything more than return a value or do a map.get is only called once.  I added 
```
    ///////////////////////////////////////////////////////////////////////////
    /// <editor-fold desc="Initialize hierarchy">
    /// Methods that initialize the hierarchy; only called once.
    ///
...
    // </editor-fold> End of methods that initialize hierarchy
```

Around these methods to make that clear.